Prepare to update bundle build by storing CMA version of CSV in this repo

joelsmith opened the pull request:
* Add tool to generate/patch a CMA-specific version of a CSV using the upstream KEDA CSV
* Add tool to detect missing or wrong CMA-specific version of any CSV after v2.7.1
* Add tool to notice improperly-formed commits so that rebasebot won't drop needed commits on the next rebase
* Add CMA-specific versions of the CSVs for versions 2.7.1, 2.8.1, 2.8.2
* Add yaml2json and json2yaml tools so that we can convert yaml to json, use jq to process data, then convert back to yaml

jrvaldes commented on Makefile:
```diff
@@ -55,6 +55,21 @@ fmt: ## Run go fmt against code.
+# utilities needed by tests
+bin/yaml2json: cmd/testutil/yaml2json/yaml2json.go
+	mkdir -p bin
+	go build $(GOGCFLAGS) -ldflags "$(LD_FLAGS)" -o bin/ "github.com/kedacore/keda-olm-operator/cmd/testutil/yaml2json"
```
Is this a valid path?  `github.com/openshift/custom-metrics-autoscaler-operator/cmd/testutil/yaml2json` maybe?

joelsmith replied:
That's a good question... The way I have my repo set up is that I cloned `github.com/kedacore/keda-olm-operator` then added a new remote of this repo. I think that as long as the CI does it the same way, it should be valid.

joelsmith replied:
Okay, once the following PR merges, I think the path should be valid. I changed the `canonical_go_repository` to be `github.com/kedacore/keda-olm-operator` so that the CI source should be set up in a location like mine.

https://github.com/openshift/release/pull/36105

jrvaldes commented on hack/cma-generate-csv.sh:
```diff
@@ -0,0 +1,59 @@
+CMA_PATCH_EOF
+
+# del() to delete array items from upstream, walk to change the deployment from keda-olm-operator to cma operator, search/replace other fields
+{ bin/yaml2json keda/${ver}/manifests/keda.v${ver}.clusterserviceversion.yaml | \
+    jq 'del(.spec.icon) | del(.spec.links) | del(.spec.maintainers) | walk(if type == "string" and . == "keda-olm-operator" then .="custom-metrics-autoscaler-operator" else . end) | .spec.replaces |= sub("keda.v"; "custom-metrics-autoscaler.v") | .metadata.annotations."alm-examples" |= sub("\"namespace\": \"keda\""; "\"namespace\": \"openshift-keda\"")';
+  echo "$cma_patch" | bin/yaml2json; } | \
+  jq --slurp 'reduce .[] as $item ({}; . * $item)' | bin/json2yaml > keda/${ver}/manifests/cma.v${ver}.clusterserviceversion.yaml.new
```
Could the transformations be performed directly in the YAML file to avoid the back-and-forth parsing?

joelsmith replied:
I wish there was a yq tool that was commonly available and powerful enough to be able to do what jq does, but if there is, I'm not aware of it.

jrvaldes replied:
[kustomize](https://github.com/kubernetes-sigs/kustomize/tree/master/examples#examples)  may be a nice alternative.

jrvaldes commented on keda/2.7.1/manifests/cma.v2.7.1.clusterserviceversion.yaml:
```diff
@@ -0,0 +1,590 @@
+apiVersion: operators.coreos.com/v1alpha1
```
Consider adding to the commit message the command used to generate the  modified files. 
For example:
```
Ran: 
  cma-generate-csv.sh 2.7.1

```

joelsmith replied:
That's a good idea. Will do!

jrvaldes commented on Makefile:
```diff
@@ -55,6 +55,21 @@ fmt: ## Run go fmt against code.
+	go build $(GOGCFLAGS) -ldflags "$(LD_FLAGS)" -o bin/ "github.com/kedacore/keda-olm-operator/cmd/testutil/yaml2json"
+bin/json2yaml: cmd/testutil/json2yaml/json2yaml.go
+	mkdir -p bin
+	go build $(GOGCFLAGS) -ldflags "$(LD_FLAGS)" -o bin/ "github.com/kedacore/keda-olm-operator/cmd/testutil/json2yaml"
```
Is this doable during the generation of the CI image used in the build process? The tool won't change very often; it may save time/resources to have the binaries present in the image.

joelsmith replied:
Good idea, I'll look into it.

joelsmith replied:
On second thought, I think that since only one CI test will be using these tools, it's fine to just build them immediately before the test. If we had several CI tests using them, then maybe we could have the CI tests create a `srcplus` image that is based on the `src` image, but build the common tools needed by various CI tests running on the `src` image, but it seems like it would increase CI testing latency and use more resources to build the intermediate image where only one test would use it. But I'll keep the idea in mind in case we add more tests that use these tools.

jrvaldes replied:
Agreed.

jrvaldes commented on hack/verify_history.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
```
How does this script correlate with `Generate/check CMA CSV files`? Consider moving the addition of this file to a separate commit (or possible PR).

joelsmith replied:
Ah yes, it's not related. I have as a goal to have few commits so that future rebases won't have to manage a lot of commits, but it's probably better to split this one out as you mention.

joelsmith replied:
I have split it into a separate commit.

jrvaldes commented on hack/cma-generate-csv.sh:
```diff
@@ -0,0 +1,59 @@
+  - name: Custom Metrics Autoscaler Documentation
+    url: https://docs.openshift.com/container-platform/latest/nodes/pods/nodes-pods-autoscaling-custom.html
+  maintainers:
+  - email: support@redhat.com
```
I lean towards using a more specialized point of contact. For example, the team mailing list.

joelsmith replied:
I'll check with our project manager to see what they suggest. We want any customers who have questions to go to support as their first point of contact, but I'm not sure if that's appropriate here.

jrvaldes replied:
I follow. Consider adding a `Help`  and/or `Reporting issues` section to the CSV's [spec.description](https://github.com/openshift/custom-metrics-autoscaler-operator/pull/14/commits/ea86006fb79d8284bf5187ac11ad8715f2dae3ed#diff-0557721b7f65b7a5ee3d5cb4853c5d5405e7d8eea542c33a0aa5027e88eb6db6R35)

jrvaldes reviewed (commented):
/lgtm

@joelsmith  Thanks for working on this. PTAL at my comments/questions.

jrvaldes commented:
/hold

joelsmith commented:
/hold cancel

@jrvaldes if you have time for a re-review, I'd appreciate it. Changes:

* Split out commit history script to its own commit
* Updated commit message to show the command run for generated CSVs.

Many thanks for your review!

jrvaldes commented:
/lgtm
